### PR TITLE
Unicode support to cp.push dir

### PR DESCRIPTION
**Pull request by rajagopalx**
### What does this PR do?
Unicode named Files and directories pushed successfully

### What issues does this PR fix?

#41018

### Previous Behavior
cp.push_dir returned False while read unicode filenames and directories.

### New Behavior
Unicode file names get successfully pushed to master 

### Tests written?

No

### CODE REVIEW

File: salt/modules/cp.py [hunk 1]
```diff
@@ -8,6 +8,9 @@
 import os
 import logging
 import fnmatch
+import sys
+reload(sys)
+sys.setdefaultencoding('utf-8')
```
Comment (cachedout) [inline salt/modules/cp.py]: This is going to set the default encoding for the entire Python process to UTF-8 despite what the system might be configured for. I don't think we want to do this.

Comment (terminalmage) [inline salt/modules/cp.py]: I agree. This is not a good fix.

File: salt/modules/cp.py [hunk 2]
```diff
@@ -900,7 +903,7 @@ def push_dir(path, glob=None, upload_path=None):
         return push(path, upload_path=upload_path)
     else:
         filelist = []
-        for root, _, files in os.walk(path):
+        for root, _, files in os.walk(ur''.join(path)):
```
Comment (terminalmage) [inline salt/modules/cp.py]: This is not valid in python 3:

```
% python3 -c 'print(ur"foo")'
  File "<string>", line 1
    print(ur"foo")
                ^
SyntaxError: invalid syntax
```

File: salt/modules/dockerng.py [hunk 3]
```diff
@@ -818,8 +818,14 @@ def _client_wrapper(attr, *args, **kwargs):
...
     func = getattr(__context__['docker.client'], attr, None)
-    if func is None:
+    if func is None or not hasattr(func, '__call__'):
         raise SaltInvocationError('Invalid client action \'{0}\''.format(attr))
+    if attr in ('push', 'pull'):
+        try:
+            # Refresh auth config from config.json
+            __context__['docker.client'].reload_config()
+        except AttributeError:
+            pass
```
Comment (terminalmage) [inline salt/modules/dockerng.py]: Why are these changes included in this pull request? They have nothing to do with ``cp.push_dir``. Not only that, they already exist in the 2016.11 branch.

**Review by terminalmage — CHANGES_REQUESTED** [review]

File: salt/modules/cp.py (hunk 2 above)
Comment (rajagopalx) [inline salt/modules/cp.py]: Yup I already mentioned this in the issue forum. we have to use `r''`  instead of `ur''`. Does we stick with python3 or is the code should be compatible with both python2 and 3 ?

File: salt/modules/dockerng.py (hunk 3 above)
Comment (rajagopalx) [inline salt/modules/dockerng.py]: I don't know how this get commited :(

File: salt/modules/cp.py (hunk 1 above)
Comment (rajagopalx) [inline salt/modules/cp.py]: I will look into other solution.

Comment (twangboy) [inline salt/modules/cp.py]: You'll have to create a _to_unicode function. Then, detect Python 2 using `six.PY2` and convert the specific variables to Unicode.

You can take a look at a PR I have for Nitrogen:
https://github.com/saltstack/salt/pull/40830/files

Comment (terminalmage) [inline salt/modules/cp.py]: Well that would need to be gated behind a check for Windows, right? And simply forcing UTF-8 wouldn't account for non-UTF-8 unicode character sets.

File: salt/modules/cp.py (hunk 2 above)
Comment (terminalmage) [inline salt/modules/cp.py]: Honestly, it would be best if we didn't try to use a str.join as a dirty hack, so the optimal solution doesn't even use this.

File: salt/modules/cp.py (hunk 1 above)
Comment (twangboy) [inline salt/modules/cp.py]: Yeah. The example above is for a Windows specific module... so there would need to be a `salt.utils.is_windows()` check as well.

The real fix here, for Windows at least, is the eventual Python 3 migration. The Unicode stuff just works. These are stop gap measures that only work in the majority of cases. You're right, it may not handle non-UTF-8 character sets, but this seemed to handle most of things I was testing with...
